### DISCUSSION TITLE
How to re-initialize a driver

### DISCUSSION THREAD

**Discussion by jeppenodgaard** [zephyrproject-rtos/zephyr#64749] [Q&A] [answered]
External ICs sometimes behave unexpectedly. This can be due to a buggy software driver, but it can also be due to hardware glitches.
Some issues can be found in chip erratas an might be possible to work around in a software driver, but some might not yet be known.
To mitigate hidden unknown drivers issues a re-initialization can be useful.
With the current driver/pm/sensor API I cannot see way to do this without creating a header with a custom, but similar reinit function.
How does one re-initialize or fix a driver that continuously responds with an error?
I feel like a common API is missing or some kind of subsystem that can provide re-initialization.

**Comment by nordicjm** [comment]
This is currently not supported

**Comment by jeppenodgaard** [comment]
Thanks @nordicjm! Do you think there is a chance Zephyr might want to support something like that?

**Reply by nordicjm** [reply, marked as answer]
Yes, there is (or were) issues about adding support for this, I did have a search earlier to post them here but wasn't able to find them

**Reply by ndrs-pst** [reply]
@jeppenodgaard I've implemented a custom function for our specific requirement.
Hope it can help.

```
extern const struct init_entry __init_POST_KERNEL_start[];
extern const struct init_entry __init_APPLICATION_start[];

int device_user_init(const struct device* dev) {
    const struct init_entry* entry;
    int rc;

    rc = -1;
    for (entry = __init_POST_KERNEL_start; entry < __init_APPLICATION_start; entry++) {
        if (entry->dev == dev) {
            rc = entry->init_fn.dev(dev);

            /* Mark device initialized.  If initialization
             * failed, record the error condition.
             */
            if (rc != 0) {
                if (rc < 0) {
                    rc = -rc;
                }

                if (rc > UINT8_MAX) {
                    rc = UINT8_MAX;
                }
                dev->state->init_res = rc;
            }
            dev->state->initialized = true;
        }
    }

    return (rc);
}
```

**Reply by yashi** [reply]
There are many duplicates, related issues, and PRs but these are open or in draft state
- https://github.com/zephyrproject-rtos/zephyr/issues/39896
- https://github.com/zephyrproject-rtos/zephyr/issues/20012
- https://github.com/zephyrproject-rtos/zephyr/pull/61986
- https://github.com/zephyrproject-rtos/zephyr/pull/64869

**Reply by ndrs-pst** [reply]
@yashi Thanks for you information
